src/vdaf/mastic.rs

```diff
@@ -0,0 +1,651 @@
+    for MasticPublicShare<VidpfWeight<T::Field>>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        mastic: &Mastic<T, P, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        MasticPublicShare::<VidpfWeight<T::Field>>::decode_with_param(
+            &(mastic.bits, mastic.vidpf.weight_parameter),
+            bytes,
+        )
+    }
+}
+
+/// Mastic input share
+///
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proofs_share: SzkProofShare<F, SEED_SIZE>,
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticInputShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.vidpf_key.encode(bytes)?;
+        self.proofs_share.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.vidpf_key.encoded_len()? + self.proofs_share.encoded_len()?)
+    }
+}
+
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        (mastic, role): &(&'a Mastic<T, P, SEED_SIZE>, usize),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        if *role > 1 {
+            return Err(CodecError::UnexpectedValue);
+        }
+        let vidpf_key = VidpfKey::decode_with_param(&(*role == 0), bytes)?;
+        let proofs_share = SzkProofShare::<T::Field, SEED_SIZE>::decode_with_param(
+            &(
+                *role == 0,
+                mastic.szk.proof_len(),
+                mastic.szk.has_joint_rand(),
+            ),
+            bytes,
+        )?;
+        Ok(Self {
+            vidpf_key,
+            proofs_share,
+        })
+    }
+}
+
+#[cfg(test)]
+impl<F: FieldElement, const SEED_SIZE: usize> PartialEq for MasticInputShare<F, SEED_SIZE> {
+    fn eq(&self, other: &MasticInputShare<F, SEED_SIZE>) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ConstantTimeEq for MasticInputShare<F, SEED_SIZE> {
+    fn ct_eq(&self, other: &MasticInputShare<F, SEED_SIZE>) -> Choice {
+        self.vidpf_key
+            .ct_eq(&other.vidpf_key)
+            .bitand(self.proofs_share.ct_eq(&other.proofs_share))
+    }
+
+    fn ct_ne(&self, other: &MasticInputShare<F, SEED_SIZE>) -> Choice {
+        self.ct_eq(other).not()
+    }
+}
+
+/// Mastic output share
+///
+/// Contains a vector of VIDPF outputs: one for each prefix.
+pub type MasticOutputShare<V> = OutputShare<V>;
+
+/// Vector of VidpfWeights to be aggregated by Mastic aggregators
+pub type MasticAggregateShare<V> = MasticOutputShare<V>;
```

Comment:
The type definition bleeds through the API boundary, so this looks a little odd.  Does this work? This would be more consistent with `Prio3`.
```suggestion
pub type MasticAggregateShare<V> = AggregateShare<V>;
```